### PR TITLE
chore(react-tree): improve ImmutableSet and ImmutableMap internal implementation

### PR DESCRIPTION
**Pull request by bsunderhus**
## New Behavior

1. converts internal implementation of `ImmutableSet` and `ImmutableMap` to classes
## Related Issue(s)

- Fixes #

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">35</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">28</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/Avatar-mount_base.html">636</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/Avatar-mount.html">652</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/Button-mount_base.html">302</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/Button-mount.html">313</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/Field-mount_base.html">1138</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/Field-mount.html">1134</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProvider-mount_base.html">735</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProvider-mount.html">703</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-mount_base.html">91</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-mount.html">90</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">35</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">28</a></td>
            <td>10</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">79</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">72</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/MakeStyles-mount_base.html">867</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/MakeStyles-mount.html">891</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/Persona-mount_base.html">1785</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/Persona-mount.html">1743</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/SpinButton-mount_base.html">1393</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/SpinButton-mount.html">1428</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/SwatchPicker-mount_base.html">1667</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32167/perf-test-react-components/SwatchPicker-mount.html">1647</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: change/@fluentui-react-tree-c6f8da33-6d54-4060-b38b-5ed11eb4d61e.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-tree-c6f8da33-6d54-4060-b38b-5ed11eb4d61e.json:1]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=32167&commitId=ef9e600322bf0d9e38d3f2c52f31f7fea12f0f7f&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 1 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask - RTL.normal.chromium.png|5| Changed|

</details>


  <details><summary><b>Tree</b> 4 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Tree.size - Dark Mode.chromium.png|1263761| Changed|
|Tree.size - High Contrast.chromium.png|1246075| Changed|
|Tree.size - RTL.chromium.png|1604087| Changed|
|Tree.size.chromium.png|1604224| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-tree/library/src/utils/ImmutableMap.ts [hunk 2]
```diff
@@ -1,77 +1,98 @@
...
-   */
-  has(key: Key): boolean;
-  /**
-   * Adds a new element with a specified key and value to the Map. If an element with the same key already exists, the element will be updated.
-   */
-  set(key: Key, value: Value): ImmutableMap<Key, Value>;
+const internalMapSymbol = Symbol('#internalMap');
+
+export class ImmutableMap<Key, Value> implements Iterable<[Key, Value]> {
+  public static empty: ImmutableMap<never, never> = new ImmutableMap(new Map<never, never>());
```
Comment (layershifter) [inline packages/react-components/react-tree/library/src/utils/ImmutableMap.ts:4]: image

Why `empty` is needed then?

**Review by layershifter — APPROVED** [review]

File: packages/react-components/react-tree/library/src/utils/ImmutableMap.ts (hunk 2 above)
Comment (bsunderhus) [reply packages/react-components/react-tree/library/src/utils/ImmutableMap.ts:4]: Good question. I guess it's redundant, I'll follow up

**Pull request merged**